File: com.teaminabox.eclipse.sortit/src/com/teaminabox/eclipse/sortit/NumericSortAction.scala
package com.teaminabox.eclipse.sortit
import java.util.regex.Matcher
import java.util.regex.Pattern

class NumericSortAction extends SortItAction {

  private val number_pattern = Pattern.compile("[-+]?([0-9]*\\.)?[0-9]+([eE][-+]?[0-9]+)?")
  override def getComparator(): (String, String) => Boolean = (a: String, b: String) => compare(a, b) < 0

  def compare(first: String, second: String): Int = getNumber(first).compareTo(getNumber(second))

  def getNumber(text: String) =
    if (text == null || text.trim().length() == 0) {
      Double.MinValue
    } else {
      findNumber(text).toDouble
    }

  def findNumber(text: String): String = {
    val m = number_pattern.matcher(text)
    if (m.find()) {
      text.substring(0, m.end())
    } else {
      text
    }
  }
}

File: com.teaminabox.eclipse.sortit/src/com/teaminabox/eclipse/sortit/NumericSortAction.java
package com.teaminabox.eclipse.sortit;
import java.util.Comparator;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class NumericSortAction extends SortItAction {

  private Pattern number_pattern = Pattern.compile("[-+]?([0-9]*\\.)?[0-9]+([eE][-+]?[0-9]+)?");
  
  public Comparator<String> getComparator() {
	  return new Comparator<String>() {
		@Override
		public int compare(String o1, String o2) {
			return compareStrings(o1, o2);
		}
	  };
  }

  private int compareStrings(String first, String second) {
	  double x = getNumber(first);
	  double y = getNumber(second);
	  if (x < y) {
		  return -1;
	  } else if (x == y) {
		  return 0;
	  }
	  return 1;
  }

  private double getNumber(String text) {
    if (text == null || text.trim().length() == 0) {
      return Double.MIN_VALUE;
    }
    return Double.valueOf(findNumber(text)).doubleValue();
  }

  private String findNumber(String text) {
    Matcher m = number_pattern.matcher(text);
    if (m.find()) {
      return text.substring(0, m.end());
    }
	return text;
  }
}
